### PR TITLE
chore(gatsby): Use Typescript internally!

### PR DESCRIPTION
**Pull request by blainekasten**
# Summary

This PR is the result of some exploration work to see what it would take to use TypeScript in gatsby source. **This does not enable TypeScript for consumers.** The goal of using TypeScript is to have the benefits of a type system to catch errors in our code paths that may not otherwise be easily found. This PR sets up the infrastructure to begin using TS and converts one file, `gatsby/commands/develop`. There is a lot of organizational and process oriented discovery work to still resolve, but they are not blockers for this.

## What's next

With this PR we enable a slow migration of changing files to TS. The community could definitely get involved here! We'll need to decide some rules on how we want to approach typing our app before we can accept PRs. Things like, sharing types and what types we allow. Our goal is to simplify the types as much as possible, complex types usually mean complex code and a refactor is warranted.

### CODE REVIEW

**Comment by moonmeister** [comment]
FYI: #13436

File: packages/gatsby/src/commands/develop.ts [hunk 1]
```diff
@@ -260,7 +280,7 @@ async function startServer(program) {
             .on(`response`, response =>
               res.writeHead(response.statusCode, response.headers)
             )
-            .on(`error`, (err, _, response) => {
+            .on(`error`, (err: Error, _: any, response) => {
```
Comment (pvdz) [inline packages/gatsby/src/commands/develop.ts]: Random but Flow has a `mixed` type. Basically means "don't know what's in it, I'm not touching it, so whatever". Does TS have this as well? Ultimately it's an `any` but the semantics are a little different and the linting can make a distinction.

Comment (pvdz) [reply packages/gatsby/src/commands/develop.ts]: (Not reviewing, just caught my eye and curious what the TS way is here, if anything)

Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: @pvdz i like the thought behind this. Unfortunately typescript doesn't. We could implement our own easily if we wanted a mixed type `declare mixed = any;`.

I did go and update this though, ideally we should be typing out everything and not using any. this was coming from the library `got` so i just added `@types/got` and now these types are properly typed out for us. So good flag. This is an example of how we should strive to not use loose types.

**Comment by blainekasten** [comment]
@moonmeister Thank you for the prior art! This is still a bit of an exploration ticket, but the core team is happy to explore this more now!

**Comment by moonmeister** [comment]
@blainekasten good to know, I think it would be huge for the project. Thanks for taking this on.

File: packages/gatsby/src/commands/develop.ts (hunk 1 above)
Comment (Austaras) [reply packages/gatsby/src/commands/develop.ts]: TS has. You can use `unknown` or `never` here and I personally recommend `never`

File: packages/gatsby/src/commands/develop.ts [hunk 2]
```diff
@@ -296,29 +319,16 @@ async function startServer(program) {
...
-        report.panic(
-          `Unable to start Gatsby on port ${program.port} as there's already a process listening on that port.`
-        )
-        return
-      }
-
-      report.panic(`There was a problem starting the development server`, err)
-    }
-  })
+  const listener = server.listen(program.port, program.host)
```
Comment (blainekasten) [inline packages/gatsby/src/commands/develop.ts]: typescript taught me that this callback function is not called with an error. But good news, we check for the port above and change the port to a new number. So this code was old and unused, but ts taught me to delete it!

File: packages/gatsby-cli/src/index.ts [hunk 3]
```diff
@@ -36,7 +34,7 @@ process.on(`unhandledRejection`, reason => {
   // across versions and crash
 
   // reason can be anything, it can be a message, an object, ANYTHING!
-  // we convert it to an error object so we don't crash on structured error validation
+  // we convtert it o an error object so we don't crash on structured error validation
   if (!(reason instanceof Error)) {
     reason = new Error(util.format(reason))
   }
```
Comment (blainekasten) [inline packages/gatsby-cli/src/index.ts]: I had to convert a second package to ensure the check-ts script could properly handle our monorepo. Good news, it does now!

File: packages/gatsby-cli/src/index.ts [hunk 4]
```diff
@@ -36,7 +34,7 @@ process.on(`unhandledRejection`, reason => {
   // across versions and crash
 
   // reason can be anything, it can be a message, an object, ANYTHING!
-  // we convert it to an error object so we don't crash on structured error validation
+  // we convtert it o an error object so we don't crash on structured error validation
```
Comment (muescha) [inline packages/gatsby-cli/src/index.ts]: this changes introduced 2 typos?
```suggestion
  // we convert it to an error object so we don't crash on structured error validation
```

File: .babelrc.js [hunk 5]
```diff
@@ -9,5 +9,16 @@ if (process.env.NODE_ENV !== `test`) {
 module.exports = {
   sourceMaps: true,
   presets: ["babel-preset-gatsby-package"],
+  overrides: [
+    {
+      test: "**/*.ts",
+      plugins: [
+        [
+          "@babel/plugin-transform-typescript",
+          { isTSX: true, allExtensions: true },
```
Comment (pvdz) [inline .babelrc.js]: Question: Is the `allExtensions`useful if we glob for `*.ts`? :)

File: packages/gatsby/src/commands/develop.ts [hunk 6]
```diff
@@ -61,7 +84,7 @@ onExit(() => {
 })
 
 const waitJobsFinished = () =>
-  new Promise((resolve, reject) => {
+  new Promise(resolve => {
```
Comment (pvdz) [inline packages/gatsby/src/commands/develop.ts]: I'm a little surprised that TS doesn't care that we don't listen to rejections here. I think node is going to crash hard for that in a future version, whenever that is.

File: packages/gatsby/src/commands/develop.ts [hunk 7]
```diff
@@ -1,51 +1,74 @@
...
+import report from "gatsby-cli/lib/reporter"
+import launchEditor from "react-dev-utils/launchEditor"
+import formatWebpackMessages from "react-dev-utils/formatWebpackMessages"
+import chalk from "chalk"
+import address from "address"
+import cors from "cors"
+import telemetry from "gatsby-telemetry"
+import WorkerPool from "../utils/worker/pool"
+import http from "http"
+import https from "https"
```
Comment (pvdz) [inline packages/gatsby/src/commands/develop.ts]: I wonder what the impact is of depending on this at build time rather than runtime.
Might be benign but it's definitely a change.

File: packages/gatsby/src/commands/develop.ts [hunk 8]
```diff
@@ -296,29 +319,16 @@ async function startServer(program) {
...
+   * If a SSL cert exists in program, use it with `createServer`.
    **/
-  let server = require(`http`).Server(app)
+  let server = program.ssl
+    ? https.createServer(program.ssl, app)
+    : new http.Server(app)
 
-  // If a SSL cert exists in program, use it with `createServer`.
-  if (program.ssl) {
-    server = require(`https`).createServer(program.ssl, app)
```
Comment (pvdz) [inline packages/gatsby/src/commands/develop.ts]: If the `require` here was deliberately inlined you could do an `await import('https')` in the parent function and then access that in a closure.
But perhaps the inlining was just something that stuck around for no reason :)

File: packages/gatsby/src/commands/develop.ts (hunk 7 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: oof, i'll measure. we only get type checking on modules (not commonjs requires). When i made the joke about shaving off 0.043s it was in reference to this, but didn't even think about the build time cost that this adds to every run.. lol

File: packages/gatsby/src/commands/develop.ts (hunk 6 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: Isn't that about the listeners more than the invocation? I thought node was going to start throwing if you did `promise.then()` and not `promise.then().catch()`

File: packages/gatsby/src/commands/develop.ts (hunk 7 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: costs `0.012ms` to import the `https` module. So I suppose it's actually a small net win if you are running in ssl, but if you are not we are adding that cost to the import. Seemingly benign for the benefit of types working

File: .babelrc.js (hunk 5 above)
Comment (blainekasten) [reply .babelrc.js]: i suppose not! kept fiddling with these options to try and get things to work and forgot what was actually important. Removed!

File: packages/gatsby/src/commands/develop.ts (hunk 8 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: I don't think there is much reason to not import in a more typical fashion.

File: packages/gatsby/src/commands/develop.ts (hunk 7 above)
Comment (pvdz) [reply packages/gatsby/src/commands/develop.ts]: Probably nothing, cool :)

File: packages/gatsby/src/commands/develop.ts (hunk 6 above)
Comment (pvdz) [reply packages/gatsby/src/commands/develop.ts]: Oh, TIL :)

File: packages/gatsby/src/commands/develop.ts [hunk 9]
```diff
@@ -1,51 +1,74 @@
...
-const { initTracer } = require(`../utils/tracer`)
-const apiRunnerNode = require(`../utils/api-runner-node`)
-const db = require(`../db`)
-const detectPortInUseAndPrompt = require(`../utils/detect-port-in-use-and-prompt`)
-const onExit = require(`signal-exit`)
-const queryUtil = require(`../query`)
-const queryWatcher = require(`../query/query-watcher`)
-const requiresWriter = require(`../bootstrap/requires-writer`)
-const {
+import url from "url"
```
Comment (wardpeet) [inline packages/gatsby/src/commands/develop.ts]: I think our imports should be `* as` style to comply with the real import. We're now relying on moduleInterop which is wrong in my opinion.

File: packages/gatsby/src/commands/develop.ts [hunk 10]
```diff
@@ -1,51 +1,74 @@
...
+import queryUtil from "../query"
+import queryWatcher from "../query/query-watcher"
+import requiresWriter from "../bootstrap/requires-writer"
+import {
   reportWebpackWarnings,
   structureWebpackErrors,
-} = require(`../utils/webpack-error-utils`)
+} from "../utils/webpack-error-utils"
+
+type Cert = {
```
Comment (wardpeet) [inline packages/gatsby/src/commands/develop.ts]: Do we want to write down guidelines when to use type or interface? I personally use interface for everything except if I need to use union, intersection, ...

File: packages/gatsby/src/commands/develop.ts [hunk 11]
```diff
@@ -461,7 +471,15 @@ module.exports = async (program: any) => {
     }
   }
 
-  function printInstructions(appName, urls, useYarn) {
+  function printInstructions(
+    appName: string,
+    urls: {
```
Comment (wardpeet) [inline packages/gatsby/src/commands/develop.ts]: Do we want to inline this or create an interface/type for it?

File: .circleci/config.yml [hunk 12]
```diff
@@ -188,6 +188,12 @@ jobs:
       - run: yarn lint:other
       - run: yarn check-repo-fields
 
+  typecheck:
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply .circleci/config.yml]: no good reason, i'll switch it!

File: packages/gatsby/src/commands/develop.ts (hunk 11 above)
Comment (pieh) [reply packages/gatsby/src/commands/develop.ts]: (IMO) I think inlining is fine. Allows type definitions to be more terse and also allow us to avoid naming things which is sometimes just hard and not strictly necessary.

I would actually default to inlining unless it cause more problems when you need to use one of those nested types?

File: packages/gatsby-cli/src/index.ts [hunk 13]
```diff
@@ -1,21 +1,19 @@
...
 
 if (useJsonLogger) {
   process.env.GATSBY_LOGGER = `json`
 }
 
-const createCli = require(`./create-cli`)
-const report = require(`./reporter`)
-const pkg = require(`../package.json`)
-const updateNotifier = require(`update-notifier`)
+import createCli from "./create-cli"
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-cli/src/index.ts]: This is not valid JS, but with us running babel this is probably not checked and ultimately transpiled to requires. We already have bunch of files with mixed `require` and `import`s that don't have imports first - like https://github.com/gatsbyjs/gatsby/blob/df1591a5d44ae0669274b422c7dac7802f82d8f4/packages/gatsby/src/bootstrap/requires-writer.js#L7-L9

File: peril/rules/merge-on-green.ts [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import { danger, peril } from "danger"
-import * as octokit from "@octokit/rest"
+import octokit from "@octokit/rest"
```
Comment (pieh) [inline peril/rules/merge-on-green.ts]: I'm not sure if changes to peril automation should be in this PR (not that we shouldn't do them at all, but it could be separate changeset).

Comment (blainekasten) [reply peril/rules/merge-on-green.ts]: when i enabled moduleInterop this was required. These files wouldn't compile anymore. I just had a discussion with ward about moduleInterop and we came to a conclusion that it is worth having on.

The only changes in any of these peril files is dropping the `* as` from the imports.

File: packages/gatsby/src/commands/develop.ts (hunk 9 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: We discussed this and think it's better to ignore the difference from commonjs and modules for a simplistic syntax.

File: peril/rules/merge-on-green.ts (hunk 14 above)
Comment (pieh) [reply peril/rules/merge-on-green.ts]: Hmmm, I'm not exactly sure how those things in Peril are run, but I don't think we control typescript config for it

I also don't have lot of context around `moduleInterop` setting - I just wonder if this change can actually break stuff depending on whatever ts config is used by Peril?

Comment (pieh) [reply peril/rules/merge-on-green.ts]: To add to previous comment - I'm relatively sure it's not us who transpile this stuff, and actually .ts files get used and Peril handles those somehow (maybe using ts-node, maybe it transpiles them as part of its internal build setup when deploying updates)

File: tsconfig.json [hunk 15]
```diff
@@ -1,8 +1,13 @@
 {
   "compilerOptions": {
-    "target": "es5",
-    "module": "commonjs",
-    "lib": ["es2017"],
-    "strict": true
-  }
+    /* Additional Checks */
+    "noUnusedLocals": true,
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply tsconfig.json]: Mostly for adoption. strict true is great but it would be so painful for our migration. Basically it turns on `noImplicityAny` which is great, except it pretty much makes every import throw. Which sucks. So any file that is imported that isn't converted would break and require us to nearly convert everything at once.

I believe a goal should be to turn strict on. I've turned on every setting that strict does _except_ that one. It sucks because that rule has so much power.

File: packages/gatsby/src/commands/develop.ts (hunk 10 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: So I've had so many debates about this. There is no longer a practical difference besides syntax with types and interfaces. I'll switch to interfaces because that's more common in the typescript world. One benefit of interfaces is the practice of prefixing with `I` makes it obvious when importing whats a type and what's code.

e.g., `import { Reporter, IReporter } from 'reporter';`

File: packages/gatsby/src/commands/develop.ts (hunk 11 above)
Comment (blainekasten) [reply packages/gatsby/src/commands/develop.ts]: i can type it

File: peril/rules/merge-on-green.ts (hunk 14 above)
Comment (pieh) [reply peril/rules/merge-on-green.ts]: Seems like Peril transpiles on execution - it will run https://github.com/danger/danger-js/blob/dab0e63a97705617c0e0017d85a9ff7129cadd5f/source/runner/runners/utils/transpiler.ts#L142-L163

whenever it needs to execute commands - it also lazily grab code it needs to execute based on https://github.com/gatsbyjs/gatsby/blob/master/peril.settings.json - so it for sure grabs ts versions and then do it's own transpilation - I'm not sure if for us it uses `typescript` or babel with typescript preset/plugin, but I'm pretty sure we don't have any custom config for it.

Maybe it would be safer for now to disable whatever was failing for those files and not go down rabbit hole too much?

File: packages/gatsby-cli/tsconfig.json [hunk 16]
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "../../tsconfig.json"
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-cli/tsconfig.json]: I don't think we can do this because the root tsconfig is used for peril.

File: packages/gatsby-cli/src/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-cli/src/index.ts]: i fixed it though. we should def be doing this right

File: packages/gatsby-cli/src/index.ts [hunk 17]
```diff
@@ -1,21 +1,19 @@
 #!/usr/bin/env node
 
-// babel-preset-env doesn't find this import if you
-// use require() with backtick strings so use the es6 syntax
 import "@babel/polyfill"
-const semver = require(`semver`)
-const util = require(`util`)
+import semver from "semver"
+import util from "util"
```
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-cli/src/index.ts]: we talked about this, sticking with the module import syntax, not the interop

[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-cli/src/index.ts]: Don't you mean sticking with automatic interop? 😛 Blaine mentioned it's easier for contributors as they don't have to think about how to import things and I think that's a great argument to do the es6 interop

File: packages/gatsby-cli/tsconfig.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-cli/tsconfig.json]: ah make sense 👍

File: packages/gatsby/src/commands/develop.ts (hunk 10 above)
Comment (wardpeet) [reply packages/gatsby/src/commands/develop.ts]: >  I've had so many debates about this

Haha, I can imagine. Honestly, I don't care, I just think something written about our way of using typescript would be great. If an eslint rule exists I would also use it to keep consistency.

File: packages/gatsby/src/commands/develop.ts (hunk 11 above)
Comment (wardpeet) [reply packages/gatsby/src/commands/develop.ts]: all good for me 👍 I like inlining for functions too unless I need to use it in 2 places.

File: peril/rules/merge-on-green.ts (hunk 14 above)
Comment (blainekasten) [reply peril/rules/merge-on-green.ts]: It must use our tsconfig though, because when I made these changes, unit tests were failing and it was all the peril ones.

File: packages/gatsby-cli/src/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (blainekasten) [reply packages/gatsby-cli/src/index.ts]: ha yeah. sorry, said that wrong!

File: peril/rules/merge-on-green.ts (hunk 14 above)
Comment (blainekasten) [reply peril/rules/merge-on-green.ts]: Okay, so this is fixed now. I left the peril files unchanged. The solution was to make the peril folder have it's own package.json jest test with it's own jest configuration that set it's tsconfig file to have esModuleInterop: false. 

I dug into the options on the tools we used and wasn't finding any other options. The only other potential option on impl is to use jest projects, but that seemed like a larger refactor.
